File: vault/desktop/src/tests/plugins/conversation.rs
//! Smoke tests for Conversation plugin DTOs

use vault::application::dtos::conversation_dto::*;

#[test]
fn test_create_conversation_request_dto() {
    let request = CreateConversationRequestDto {
        title: "Research Discussion".to_string(),
        model_name: "gpt-4".to_string(),
        system_prompt: Some("You are a helpful assistant.".to_string()),
    };

    assert_eq!(request.title, "Research Discussion");
    assert_eq!(request.model_name, "gpt-4");
    assert!(request.system_prompt.is_some());
}

#[test]
fn test_create_conversation_request_minimal() {
    let request = CreateConversationRequestDto {
        title: "Quick Chat".to_string(),
        model_name: "gpt-3.5-turbo".to_string(),
        system_prompt: None,
    };

    assert_eq!(request.title, "Quick Chat");
    assert!(request.system_prompt.is_none());
}

#[test]
fn test_list_conversations_query() {
    let query = ListConversationsQuery {
        limit: Some(20),
        offset: Some(0),
    };

    assert_eq!(query.limit.unwrap(), 20);
    assert_eq!(query.offset.unwrap(), 0);
}

#[test]
fn test_list_conversations_query_defaults() {
    let query = ListConversationsQuery {
        limit: None,
        offset: None,
    };

    assert!(query.limit.is_none());
    assert!(query.offset.is_none());
}

#[test]
fn test_get_conversation_request_dto() {
    let request = GetConversationRequestDto {
        conversation_id: "conv-123".to_string(),
    };

    assert_eq!(request.conversation_id, "conv-123");
}

#[test]
fn test_rename_conversation_request_dto() {
    let request = RenameConversationRequestDto {
        conversation_id: "conv-456".to_string(),
        new_title: "Updated Title".to_string(),
    };

    assert_eq!(request.conversation_id, "conv-456");
    assert_eq!(request.new_title, "Updated Title");
}

#[test]
fn test_delete_conversation_request_dto() {
    let request = DeleteConversationRequestDto {
        conversation_id: "conv-789".to_string(),
    };

    assert_eq!(request.conversation_id, "conv-789");
}

#[test]
fn test_conversation_dto_complete() {
    let dto = ConversationDto {
        id: "conv-111".to_string(),
        title: "Test Conversation".to_string(),
        model_name: "gpt-4".to_string(),
        system_prompt: Some("You are helpful".to_string()),
        created_at: "2024-01-01T00:00:00Z".to_string(),
        updated_at: "2024-01-01T00:00:00Z".to_string(),
        message_count: 10,
        total_tokens: 500,
        space_id: Some("space_general".to_string()),
        is_saved: Some(false),
        is_bookmarked: Some(false),
        is_pinned: Some(false),
        is_archived: Some(false),
        saved_at: None,
        bookmarked_at: None,
        pinned_at: None,
        archived_at: None,
        last_message_preview: Some("Most recent content".to_string()),
    };

    assert_eq!(dto.id, "conv-111");
    assert_eq!(dto.message_count, 10);
}

#[test]
fn test_message_dto_creation() {
    let dto = MessageDto {
        id: "msg-222".to_string(),
        conversation_id: "conv-111".to_string(),
        role: "user".to_string(),
        content: "Hello".to_string(),
        status: "completed".to_string(),
        tokens: 5,
        created_at: "2024-01-01T00:00:00Z".to_string(),
        metadata: None,
    };

    assert_eq!(dto.role, "user");
    assert_eq!(dto.tokens, 5);
}

File: src/app/websrc/types/api/conversation.ts
/**
 * Conversation API Types
 *
 * Type definitions for conversation management operations.
 * These types match the Rust backend structures from application/dtos/modules/conversation_dto.rs
 */

/**
 * Conversation representation with metadata and usage statistics.
 */
export interface ConversationDto {
  /** Conversation identifier (UUID) */
  id: string;
  /** Conversation title */
  title: string;
  /** LLM model name (e.g., "claude-sonnet-4-5-20250929") */
  modelName: string;
  /** Optional system prompt */
  systemPrompt: string | null;
  /** Creation timestamp (ISO 8601) */
  createdAt: string;
  /** Last update timestamp (ISO 8601) */
  updatedAt: string;
  /** Number of messages in conversation */
  messageCount: number;
  /** Total tokens used across all messages */
  totalTokens: number;
  /** Space/environment identifier that owns this conversation */
  spaceId?: string | null;
  /** Whether this conversation is saved */
  isSaved?: boolean | null;
  /** Whether this conversation is bookmarked */
  isBookmarked?: boolean | null;
  /** Whether this conversation is pinned */
  isPinned?: boolean | null;
  /** Whether this conversation is archived */
  isArchived?: boolean | null;
  /** Timestamp when saved (ISO 8601) */
  savedAt?: string | null;
  /** Timestamp when bookmarked (ISO 8601) */
  bookmarkedAt?: string | null;
  /** Timestamp when pinned (ISO 8601) */
  pinnedAt?: string | null;
  /** Timestamp when archived (ISO 8601) */
  archivedAt?: string | null;
  /** Optional preview of latest message */
  lastMessagePreview?: string | null;
}

/**
 * Conversation message with role and content.
 */
export interface MessageDto {
  /** Message identifier (UUID) */
  id: string;
  /** Conversation identifier this message belongs to */
  conversationId: string;
  /** Message role (user, assistant, system) */
  role: 'user' | 'assistant' | 'system';
  /** Message content */
  content: string;
  /** Token count for this message */
  tokens: number;
  /** Creation timestamp (ISO 8601) */
  createdAt: string;
  /** Optional metadata (JSON string) */
  metadata: string | null;
  /** Message status (pending/completed/failed) */
  status: 'pending' | 'completed' | 'failed';
}

// Request DTOs

/**
 * Request to create a new conversation.
 */
export interface CreateConversationRequest {
  /** Conversation title */
  title: string;
  /** LLM model name */
  modelName: string;
  /** Optional system prompt */
  systemPrompt?: string;
}

/**
 * Request to get a conversation by ID.
 */
export interface GetConversationRequest {
  /** Conversation identifier */
  conversationId: string;
}

/**
 * Request to get conversation messages.
 */
export interface GetConversationMessagesRequest {
  /** Conversation identifier */
  conversationId: string;
}

/**
 * Request to rename a conversation.
 */
export interface RenameConversationRequest {
  /** Conversation identifier */
  conversationId: string;
  /** New title */
  newTitle: string;
}

/**
 * Request to delete a conversation.
 */
export interface DeleteConversationRequest {
  /** Conversation identifier */
  conversationId: string;
}

/**
 * Query parameters for listing conversations.
 */
export interface ListConversationsQuery {
  /** Maximum number of conversations to return */
  limit?: number;
  /** Offset for pagination */
  offset?: number;
}

// Response DTOs

/**
 * Response from creating a conversation.
 */
export interface CreateConversationResponse {
  /** The created conversation */
  conversation: ConversationDto;
  /** Status message */
  status: string;
}

/**
 * Response from getting a conversation.
 */
export interface GetConversationResponse {
  /** The conversation (null if not found) */
  conversation: ConversationDto | null;
}

/**
 * Response from getting conversation messages.
 */
export interface GetConversationMessagesResponse {
  /** List of messages in chronological order */
  messages: MessageDto[];
  /** Total count */
  total: number;
}

/**
 * Response from renaming a conversation.
 */
export interface RenameConversationResponse {
  /** Status message */
  status: string;
}

/**
 * Response from deleting a conversation.
 */
export interface DeleteConversationResponse {
  /** Status message */
  status: string;
}

/**
 * Response from listing conversations.
 */
export interface ListConversationsResponse {
  /** List of conversations */
  conversations: ConversationDto[];
  /** Total count (for pagination) */
  total: number;
}

/**
 * Request to synthesize multiple journal conversations into one structured summary.
 */
export interface SynthesizeJournalEntriesRequest {
  /** Ordered conversation IDs to include in synthesis */
  conversationIds: string[];
  /** Scope label for synthesis metadata */
  scope?: 'current' | 'deck' | 'pinned' | string;
  /** Optional cap for backend processing */
  maxEntries?: number;
}

/**
 * Response from journal synthesis command.
 */
export interface SynthesizeJournalEntriesResponse {
  /** Final synthesis markdown */
  synthesis: string;
  /** Scope echoed by backend */
  scope: string;
  /** Entries actually synthesized */
  entryCount: number;
  /** Number of map chunks processed */
  chunkCount: number;
  /** Conversation IDs used in synthesis */
  conversationIds: string[];
}

/**
 * Conversation space/environment DTO.
 */
export interface ConversationSpaceDto {
  id: string;
  name: string;
  description: string | null;
  icon: string | null;
  accentColor: string | null;
  spacePrompt: string | null;
  defaultModelName: string | null;
  toolPreferencesJson: string | null;
  isArchived: boolean;
  sortOrder: number;
  createdAt: string;
  updatedAt: string;
}

export interface ConversationJournalDto {
  id: string;
  name: string;
  description: string | null;
  icon: string | null;
  accentColor: string | null;
  spacePrompt: string | null;
  defaultModelName: string | null;
  toolPreferencesJson: string | null;
  isArchived: boolean;
  sortOrder: number;
  createdAt: string;
  updatedAt: string;
}

export interface CreateConversationSpaceRequest {
  name: string;
  description?: string | null;
  icon?: string | null;
  accentColor?: string | null;
  spacePrompt?: string | null;
  defaultModelName?: string | null;
  toolPreferencesJson?: string | null;
}

export interface CreateConversationJournalRequest {
  name: string;
  description?: string | null;
  icon?: string | null;
  accentColor?: string | null;
  spacePrompt?: string | null;
  defaultModelName?: string | null;
  toolPreferencesJson?: string | null;
}

export interface UpdateConversationSpaceRequest {
  spaceId: string;
  name?: string;
  description?: string | null;
  icon?: string | null;
  accentColor?: string | null;
  spacePrompt?: string | null;
  defaultModelName?: string | null;
  toolPreferencesJson?: string | null;
  isArchived?: boolean;
  sortOrder?: number;
}

export interface UpdateConversationJournalRequest {
  journalId: string;
  name?: string;
  description?: string | null;
  icon?: string | null;
  accentColor?: string | null;
  spacePrompt?: string | null;
  defaultModelName?: string | null;
  toolPreferencesJson?: string | null;
  isArchived?: boolean;
  sortOrder?: number;
}

export interface ArchiveConversationSpaceRequest {
  spaceId: string;
  archived: boolean;
}

export interface ArchiveConversationJournalRequest {
  journalId: string;
  archived: boolean;
}

export interface DeleteConversationJournalRequest {
  journalId: string;
}

export interface MoveConversationToSpaceRequest {
  conversationId: string;
  spaceId: string;
}

export interface AddConversationToJournalRequest {
  journalSpaceId: string;
  conversationId: string;
}

export interface RemoveConversationFromJournalRequest {
  journalSpaceId: string;
  conversationId: string;
}

export interface SetConversationStateRequest {
  conversationId: string;
  value: boolean;
}

export interface ListConversationsExplorerQuery {
  spaceId?: string;
  query?: string;
  savedOnly?: boolean;
  bookmarkedOnly?: boolean;
  pinnedOnly?: boolean;
  hasMessageBookmarks?: boolean;
  includeArchived?: boolean;
  limit?: number;
  offset?: number;
}

export interface ListJournalConversationsQuery {
  journalSpaceId: string;
  query?: string;
  includeArchived?: boolean;
  limit?: number;
  offset?: number;
}

export interface ConversationMessageBookmarkDto {
  id: string;
  conversationId: string;
  conversationTitle: string;
  spaceId: string;
  messageId: string;
  messageRole: 'user' | 'assistant' | 'system';
  messagePreview: string;
  title: string | null;
  note: string | null;
  createdAt: string;
}

export interface BookmarkConversationMessageRequest {
  conversationId: string;
  messageId: string;
  title?: string | null;
  note?: string | null;
}

export interface UnbookmarkConversationMessageRequest {
  conversationId: string;
  messageId: string;
}

export interface DeleteConversationMessageRequest {
  conversationId: string;
  messageId: string;
}

export interface ListMessageBookmarksQuery {
  conversationId?: string;
  query?: string;
  limit?: number;
  offset?: number;
}

export interface ListMessageBookmarksResponse {
  bookmarks: ConversationMessageBookmarkDto[];
  total: number;
}

export interface ConversationLinkedDocumentDto {
  documentId: string;
  fileName: string;
  filePath: string;
  fileType: string;
  category: string;
  indexedAt: string;
  lastReferencedAt: string;
  referenceCount: number;
}

export interface ConversationWebSourceDto {
  id: string;
  url: string;
  normalizedUrl: string;
  title?: string | null;
  excerpt?: string | null;
  relevanceScore?: number | null;
  addedAt: string;
}

export interface DocumentSpaceMembershipDto {
  spaceId: string;
  spaceName: string;
  isArchived: boolean;
  createdAt?: string | null;
}

export interface ConversationSpaceMemberDto {
  spaceId: string;
  memberId: string;
  displayName: string;
  email?: string | null;
  avatarUrl?: string | null;
  role: 'owner' | 'editor' | 'viewer';
  createdAt: string;
  updatedAt: string;
}

export interface UpsertConversationSpaceMemberRequest {
  spaceId: string;
  memberId: string;
  displayName?: string;
  email?: string | null;
  avatarUrl?: string | null;
  role: 'owner' | 'editor' | 'viewer';
}

export interface RemoveConversationSpaceMemberRequest {
  spaceId: string;
  memberId: string;
}
